uitools/import/Esri/ArcGISRuntime/Toolkit/Controller/FloorFilterController.qml

```diff
@@ -0,0 +1,686 @@
+
+    onFloorManagerChanged: {
+        internal.onFloorManagerLoaded();
+    }
+
+    /*!
+      \brief Load the \l floorManager associated with the \l geoModel and populate the sites.
+      \sa populateSites
+    */
+    property Connections floorManagerConn: Connections {
+        target: floorManager
+        function onLoadStatusChanged() {
+            internal.onFloorManagerLoaded();
+        }
+    }
+
+    /*!
+      \qmlproperty bool selectedSiteRespected
+      \brief Repopulates the \l facilities model when toggled false.
+      When the selected site is respected, the facilities list is limited to the facilities which match the selected site.
+      If the selected site is not respected, then facilities will contain all facilities from all sites, regardless of the selected site.
+      Default is \c {true}.
+    */
+    property bool selectedSiteRespected: true
+
+    onSelectedSiteRespectedChanged: {
+        if (!selectedSiteRespected) {
+            populateAllFacilities();
+        }
+    }
+
+    /*!
+      \qmlproperty string selectedLevelId
+      \brief Stores the current selected level Id.
+      Default is \c {""}.
+    */
+    readonly property alias selectedLevelId: internal.selectedLevelId
+
+    /*!
+      \brief Sets the \a levelId as \l selectedLevelId.
+      This will change the current visibile level on the \c GeoView.
+      If \a levelId is empty string, current selected level is deseleted and its visibility is set to false.
+      Resets the previous level visibility. \l updateLevelsMode modifies the behavior of selecting level.
+
+    */
+    function setSelectedLevelId(levelId) {
+        internal.selectedLevelId = levelId;
+    }
+
+    /*!
+      \qmlproperty string selectedFacilityId
+      \brief Stores the current selected facility Id.
+      Default is \c {""}.
+    */
+    readonly property alias selectedFacilityId: internal.selectedFacilityId
+
+    /*!
+      \brief Sets the \a facilityId as \l selectedFacilityId.
+      This will load the \l levels with the levels withing the facility and
+      reset by selecting \l selectedLevelId with level matching vertical ordering == 0.
+      Set empty string to reset the facility.
+      \sa setSelectedLevelId
+    */
+    function setSelectedFacilityId(facilityId) {
+        // avoid findind null idx, immediately return if empty string
+        if(facilityId === ""){
+            internal.selectedFacilityId = facilityId;
+            return;
+        }
+
+        let idx = findElementIdxById(facilityId,
+                                     FloorFilterController.TypeElement.Facility);
+        if (idx == null) {
+            console.error("not found facility");
+            internal.selectedFacilityId = "";
+            return;
+        }
+
+        // check that the current site is the selected facility selected's parent site. Otherwise the click came from a populateAllFacilities
+        var facility = floorManager.facilities[idx];
```

Comment:
At the moment is not possible to return an `idx` that will be outofbounds, because `findElementIdxById` is iterating over the same container `floorManager.facilities`. Do you recommend to protect from future possible changes with a outofbound check?

Reply:
your call. doesn't do any harm to be future proof